**CNNs (Convolutional Neural Networks) for MRI**

In [1]:
import os
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [2]:
train_data_dir = './train'
test_data_dir = './test'

In [3]:
class_names = ['no_tumor', 'pituitary_tumor', 'meningioma_tumor','glioma_tumor']

In [4]:
datagen = ImageDataGenerator(rescale=1./255)

batch_size = 32 #number of training examples in one forward/backward pass. higher batch size = more memory space needed

train_generator = datagen.flow_from_directory(
        train_data_dir,
        target_size=(150, 150),
        batch_size=batch_size,
        class_mode='categorical')

validation_generator = datagen.flow_from_directory(
        test_data_dir,
        target_size=(150, 150),
        batch_size=batch_size,
        class_mode='categorical')

Found 2870 images belonging to 4 classes.
Found 394 images belonging to 4 classes.


In [5]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(16, (3,3), activation='relu', input_shape=(150, 150, 3)),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(32, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Conv2D(64, (3,3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2,2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(512, activation='relu'),
    tf.keras.layers.Dense(4, activation='softmax')
])


In [6]:
model.compile(loss='categorical_crossentropy',
              optimizer=tf.keras.optimizers.Adam(),
              metrics=['accuracy'])

history = model.fit(train_generator,
                    steps_per_epoch=train_generator.n//train_generator.batch_size,
                    epochs=10,
                    validation_data=validation_generator,
                    validation_steps=validation_generator.n//validation_generator.batch_size)

Epoch 1/10
89/89 [==============================] - 44s 484ms/step - loss: 0.9594 - accuracy: 0.6131 - val_loss: 1.8877 - val_accuracy: 0.3776
Epoch 2/10
89/89 [==============================] - 15s 167ms/step - loss: 0.5098 - accuracy: 0.7963 - val_loss: 1.6610 - val_accuracy: 0.5547
Epoch 3/10
89/89 [==============================] - 15s 173ms/step - loss: 0.3246 - accuracy: 0.8753 - val_loss: 1.6658 - val_accuracy: 0.6693
Epoch 4/10
89/89 [==============================] - 18s 196ms/step - loss: 0.1985 - accuracy: 0.9341 - val_loss: 2.3707 - val_accuracy: 0.6615
Epoch 5/10
89/89 [==============================] - 16s 174ms/step - loss: 0.1119 - accuracy: 0.9612 - val_loss: 2.4610 - val_accuracy: 0.6901
Epoch 6/10
89/89 [==============================] - 20s 219ms/step - loss: 0.0512 - accuracy: 0.9838 - val_loss: 3.6846 - val_accuracy: 0.7240
Epoch 7/10
89/89 [==============================] - 19s 208ms/step - loss: 0.0490 - accuracy: 0.9831 - val_loss: 3.2802 - val_accuracy: 0.6927

In [8]:
test_generator = datagen.flow_from_directory(
        test_data_dir,
        target_size=(150, 150),
        batch_size=batch_size,
        class_mode='categorical',
        shuffle=False)

model.evaluate(test_generator)

Found 394 images belonging to 4 classes.
13/13 [==============================] - 1s 80ms/step - loss: 4.0390 - accuracy: 0.7411


[4.039004802703857, 0.7411167621612549]

The model correctly identified 74.11% of images in the dataset. The model's loss value on the dataset was 4.04. Loss value is equal to the average difference between predicted output of the model and actual output. 